### PR TITLE
refactor(table-duplicate-name): use virtual node

### PR DESCRIPTION
**Pull request by dbowling**
- use virtual node

references: #3473

### CODE REVIEW

**Review by WilcoFiers — CHANGES_REQUESTED** [review]

File: lib/checks/tables/same-caption-summary-evaluate.js [hunk 1]
```diff
@@ -1,12 +1,34 @@
...
+      ? virtualNode
+      : getNodeFromTree(node);
+
+  if (virtualNode && virtualNode.props.nodeName !== 'table') {
+    return undefined;
+  }
+
+  if (virtualNode.children === undefined) {
+    // undefined in this check means pass
+    return false;
```
[the remark this answers is not in the document]
Comment (dbowling) [reply lib/checks/tables/same-caption-summary-evaluate.js:19]: This change will be here with the next push.

File: lib/checks/tables/same-caption-summary-evaluate.js [hunk 2]
```diff
@@ -1,12 +1,34 @@
...
+  var captionNode = virtualNode.children.find(isCaptionNode);
+  var caption = captionNode ? sanitize(subtreeText(captionNode)) : false;
 
-function sameCaptionSummaryEvaluate(node) {
-  // passing node.caption to accessibleText instead of using
-  // the logic in accessibleTextVirtual on virtualNode
   return (
-    !!(node.summary && node.caption) &&
-    node.summary.toLowerCase() === accessibleText(node.caption).toLowerCase()
+    !!(summary && caption) &&
```
[the remark this answers is not in the document]
Comment (dbowling) [reply lib/checks/tables/same-caption-summary-evaluate.js:27]: I had simplified it from the original code, but in reality this should be deleted. The `!!` just does a type conversion that `(summary && caption)` already does.

A commit is coming up that removes this line.

File: test/integration/virtual-rules/table-duplicate-name.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
+describe('table-duplicate-name virtual-rule', function () {
+  it('should pass on table element', function () {
+    var tableNode = new axe.SerialVirtualNode({
+      nodeName: 'table'
+    });
+
+    var results = axe.runVirtualRule('table-duplicate-name', tableNode);
+
+    assert.lengthOf(results.passes, 1);
```
[the remark this answers is not in the document]
Comment (dbowling) [reply test/integration/virtual-rules/table-duplicate-name.js:24]: The diff tied your comment to a different place after the update, but this has been addressed. There is now a test for missing nodes (which incomplete) and empty children (which passes.)

I am marking this resolved.

File: lib/checks/tables/same-caption-summary-evaluate.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dbowling) [reply lib/checks/tables/same-caption-summary-evaluate.js]: This has been updated to return undefined and an "incomplete" message has been added.

File: lib/checks/tables/same-caption-summary-evaluate.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbowling) [reply lib/checks/tables/same-caption-summary-evaluate.js]: It's in a separate statement now, and I converted it to a more commonly used comparison than the double-bang it had before.

File: lib/checks/tables/same-caption-summary-evaluate.js [hunk 4]
```diff
@@ -1,12 +1,27 @@
...
-    node.summary.toLowerCase() === accessibleText(node.caption).toLowerCase()
-  );
-}
+import { sanitize, subtreeText } from '../../commons/text';
 
 export default sameCaptionSummaryEvaluate;
+
+function sameCaptionSummaryEvaluate(node, options, virtualNode) {
+  if (virtualNode.props.nodeName !== 'table') {
+    return undefined;
```
Comment (WilcoFiers) [inline lib/checks/tables/same-caption-summary-evaluate.js:7]: As I said yesterday (https://github.com/dequelabs/axe-core/pull/3604#discussion_r945902734) this isn't necessary. We don't need to check that the node matches the selector in the check. Axe has more than enough tests to make sure of that.

File: lib/checks/tables/same-caption-summary.json [hunk 5]
```diff
@@ -5,7 +5,8 @@
     "impact": "minor",
     "messages": {
       "pass": "Content of summary attribute and <caption> are not duplicated",
-      "fail": "Content of summary attribute and <caption> element are identical"
+      "fail": "Content of summary attribute and <caption> element are identical",
+      "incomplete": "Unable to determine if <table> element has children."
```
Comment (WilcoFiers) [inline lib/checks/tables/same-caption-summary.json:9]:
```suggestion
      "incomplete": "Unable to determine if <table> element has a caption."
```

**Review by WilcoFiers — CHANGES_REQUESTED** [review]

File: lib/checks/tables/same-caption-summary-evaluate.js (hunk 4 above)
Comment (dbowling) [reply lib/checks/tables/same-caption-summary-evaluate.js]: Thank you Wilco, makes sense re: selector responsibility for this.

File: lib/checks/tables/same-caption-summary.json (hunk 5 above)
Comment (dbowling) [reply lib/checks/tables/same-caption-summary.json]: Made the change, and took out the period which was not present in other messages.

**Review by WilcoFiers — APPROVED** [review]

**Pull request merged**